File: sdk/runanywhere-swift/Sources/RunAnywhere/Infrastructure/Logging/SentryDestination.swift
//
//  SentryDestination.swift
//  RunAnywhere SDK
//
//  Log destination that sends logs to Sentry for error tracking
//

import Sentry

/// Log destination that sends warning+ logs to Sentry
public final class SentryDestination: LogDestination, @unchecked Sendable {

    // MARK: - LogDestination

    public static let destinationID = "com.runanywhere.logging.sentry"

    public let identifier: String = SentryDestination.destinationID

    public var isAvailable: Bool {
        SentryManager.shared.isInitialized
    }

    /// Only send warning level and above to Sentry
    private let minSentryLevel: RALogLevel = .warning

    public init() {}

    // MARK: - LogDestination Operations

    public func write(_ entry: RALogEntry) {
        guard entry.level >= minSentryLevel, isAvailable else { return }

        // Add as breadcrumb for context trail
        addBreadcrumb(for: entry)

        // For error and fault levels, capture as Sentry event
        if entry.level >= .error {
            captureEvent(for: entry)
        }
    }

    public func flush() {
        guard isAvailable else { return }
        SentrySDK.flush(timeout: 2.0)
    }

    // MARK: - Private Helpers

    private func addBreadcrumb(for entry: RALogEntry) {
        let breadcrumb = Breadcrumb(level: convertToSentryLevel(entry.level), category: entry.category)
        breadcrumb.message = entry.message
        breadcrumb.timestamp = Self.timestamp(from: entry)

        if !entry.metadata.isEmpty {
            breadcrumb.data = entry.metadata
        }

        SentrySDK.addBreadcrumb(breadcrumb)
    }

    private func captureEvent(for entry: RALogEntry) {
        let event = Event(level: convertToSentryLevel(entry.level))
        event.message = SentryMessage(formatted: entry.message)
        event.timestamp = Self.timestamp(from: entry)
        event.tags = [
            "category": entry.category,
            "log_level": "\(entry.level)"
        ]

        // Device metadata is folded into `entry.metadata` SDK-side (RALogEntry
        // has no native device field): keys device_model / os_version / platform.
        if !entry.metadata.isEmpty {
            event.extra = entry.metadata
        }

        SentrySDK.capture(event: event)
    }

    /// Convert the proto's `timestampUnixMs: Int64` back to a `Date`.
    private static func timestamp(from entry: RALogEntry) -> Date {
        Date(timeIntervalSince1970: Double(entry.timestampUnixMs) / 1000)
    }

    private func convertToSentryLevel(_ level: RALogLevel) -> SentryLevel {
        switch level {
        case .trace: return .debug
        case .debug: return .debug
        case .info: return .info
        case .warning: return .warning
        case .error: return .error
        case .fatal: return .fatal
        case .UNRECOGNIZED: return .info
        }
    }
}

File: sdk/runanywhere-kotlin/src/main/kotlin/com/runanywhere/sdk/infrastructure/logging/SentryDestination.kt
/*
 * Copyright 2026 RunAnywhere SDK
 * SPDX-License-Identifier: Apache-2.0
 *
 * Log destination that sends logs to Sentry for error tracking.
 * Matches iOS SDK's SentryDestination.swift.
 */

package com.runanywhere.sdk.infrastructure.logging

import ai.runanywhere.proto.v1.LogEntry
import ai.runanywhere.proto.v1.LogLevel
import io.sentry.Breadcrumb
import io.sentry.Sentry
import io.sentry.SentryEvent
import io.sentry.SentryLevel
import io.sentry.protocol.Message
import java.util.Date

/**
 * Log destination that sends warning+ logs to Sentry.
 *
 * - Warning level: Added as breadcrumbs for context trail
 * - Error/Fault level: Captured as Sentry events
 */
class SentryDestination : LogDestination {
    companion object {
        const val DESTINATION_ID = "com.runanywhere.logging.sentry"
    }

    /**
     * Unique identifier for this destination.
     */
    override val identifier: String = DESTINATION_ID

    /**
     * Whether this destination is available for writing.
     */
    override val isAvailable: Boolean
        get() = SentryManager.isInitialized

    /**
     * Only send warning level and above to Sentry.
     */
    private val minSentryLevel: LogLevel = LogLevel.LOG_LEVEL_WARNING

    // Log destination operations

    /**
     * Write a log entry to Sentry.
     *
     * @param entry The log entry to write
     */
    override fun write(entry: LogEntry) {
        // proto LogLevel: larger value = more severe (TRACE=0, DEBUG=1,
        // INFO=2, WARNING=3, ERROR=4, FATAL=5). Send to Sentry iff entry
        // severity is at or above the min Sentry level.
        if (!isAvailable || entry.level.value < minSentryLevel.value) {
            return
        }

        // Add as breadcrumb for context trail
        addBreadcrumb(entry)

        // ERROR + FATAL capture as Sentry events; WARNING stays a breadcrumb.
        if (entry.level.value >= LogLevel.LOG_LEVEL_ERROR.value) {
            captureEvent(entry)
        }
    }

    /**
     * Flush any buffered entries.
     */
    override fun flush() {
        if (!isAvailable) return
        SentryManager.flush()
    }

    // Private helpers

    /**
     * Add a breadcrumb for context trail.
     */
    private fun addBreadcrumb(entry: LogEntry) {
        val timestamp = Date(entry.timestamp_unix_ms)
        val breadcrumb =
            Breadcrumb(timestamp).apply {
                category = entry.category
                message = entry.message
                level = convertToSentryLevel(entry.level)
                entry.metadata.forEach { (key, value) ->
                    setData(key, value)
                }
            }

        Sentry.addBreadcrumb(breadcrumb)
    }

    /**
     * Capture an error event in Sentry.
     */
    private fun captureEvent(entry: LogEntry) {
        val timestamp = Date(entry.timestamp_unix_ms)
        val event =
            SentryEvent(timestamp).apply {
                level = convertToSentryLevel(entry.level)
                message =
                    Message().apply {
                        formatted = entry.message
                    }

                // Add tags
                setTag("category", entry.category)
                setTag("log_level", entry.level.toString())

                // Add metadata as extras
                entry.metadata.forEach { (key, value) ->
                    setExtra(key, value)
                }

                // Add model info if present (proto: "" / 0 means unset)
                entry.model_id.takeIf { it.isNotEmpty() }?.let { setExtra("model_id", it) }
                entry.framework.takeIf { it.isNotEmpty() }?.let { setExtra("framework", it) }
                entry.error_code.takeIf { it != 0 }?.let { setExtra("error_code", it) }

                // Add source location if present
                entry.file_.takeIf { it.isNotEmpty() }?.let { setExtra("source_file", it) }
                entry.line.takeIf { it != 0 }?.let { setExtra("source_line", it) }
                entry.function.takeIf { it.isNotEmpty() }?.let { setExtra("source_function", it) }
            }

        Sentry.captureEvent(event)
    }

    /**
     * Convert SDK LogLevel to Sentry level.
     */
    private fun convertToSentryLevel(level: LogLevel): SentryLevel {
        return when (level) {
            LogLevel.LOG_LEVEL_TRACE -> SentryLevel.DEBUG
            LogLevel.LOG_LEVEL_DEBUG -> SentryLevel.DEBUG
            LogLevel.LOG_LEVEL_INFO -> SentryLevel.INFO
            LogLevel.LOG_LEVEL_WARNING -> SentryLevel.WARNING
            LogLevel.LOG_LEVEL_ERROR -> SentryLevel.ERROR
            LogLevel.LOG_LEVEL_FATAL -> SentryLevel.FATAL
        }
    }
}
